### PR TITLE
mlxrunner: implement load progress reporting and stall detection

### PR DESCRIPTION
**Pull request by dhiltgen**
Implements the pre-0.30 progress-based stall detection to avoid possibly canceling slow/long loads that exceed the default timeout.

Three MLX-specific wrinkles. The listener now binds before the load, since the runner previously served nothing until the model was resident. Weights load lazily, so the single mlx.Eval over every array is split into chunks to give the fraction something to move on. And Ping now fails unless the runner reports ready, since the scheduler reads MemorySize before the load ends and would otherwise replace its own estimate with a partial one.

Progress covers dense models fully; architectures that fuse expert weights materialize most of the model inside LoadWeights and report only the rest.

### CODE REVIEW

**Review by jessegross — COMMENTED** [review]
I think another possible approach is to actually read the bytes for tensor loads ourselves in Go and count them as they go through. MLX lets you supply a reader to handle file I/O. We can initially set it up to just mirror MLX's threading design.

If we did it that way, we wouldn't need to break up the actual evals and it would be robust to changes in our loading (for example, the change that I just made to eval the loads on the CPU first). I also think that we will need to go this way anyways to get more control over loading behavior - we previously received feedback that reading in parallel from the same FD as MLX does causes problems for some network filesystems.

File: x/mlxrunner/mlx/io_reader.go [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+	if r.closed.Load() {
+		if err := file.Close(); err != nil {
+			r.setErr(err)
+		}
+		return
+	}
+	r.idle = append(r.idle, file)
+}
+
+func (r *fileIOReader) setErr(err error) {
```
Comment (jessegross) [inline x/mlxrunner/mlx/io_reader.go]: With the current mlx-c a short read throws inside MLX and fails the eval, the same as the native loader. The error latch, `Err`, `good` and `errMu` can go.

File: x/mlxrunner/mlx/io_reader.go [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+			r.progress(int64(res.read))
+		}
+		if res.err != nil {
+			r.setErr(res.err)
+		}
+	}
+	return total
+}
+
+func (r *fileIOReader) acquireFile() (*os.File, error) {
```
Comment (jessegross) [inline x/mlxrunner/mlx/io_reader.go]: The pool hands whichever description is idle to whichever 32 MiB batch is next, so a single description still sees strided offsets. The point of separate descriptions is a sequential stream per description for readahead: split a large read into four contiguous ranges and pin range k to description k.

File: x/mlxrunner/mlx/io.go [hunk 3]
```diff
@@ -26,25 +26,58 @@ func loadSafetensorsStream() C.mlx_stream {
...
-
 	stream := loadSafetensorsStream()
 	if err := mlxError(stream); err != nil {
 		return nil, err
 	}
 	defer freeStream(stream)
 
-	if err := mlxError(C.mlx_load_safetensors(&arrays, &metadata, cPath, stream)); err != nil {
-		return nil, fmt.Errorf("failed to load safetensors %s: %w", path, err)
+	if progress == nil {
```
Comment (jessegross) [inline x/mlxrunner/mlx/io.go]: Nothing should use the native loader anymore. One loader that always goes through the Go reader, with a nil progress meaning no callback; quantize passes nil. That removes `LoadSafetensorsNative`, `LoadSafetensorsWithProgress`, `Load`, `load` and the runner's `loadSafetensors`.

File: x/mlxrunner/mlx/io_reader.go [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+
+// Match MLX's native ParallelFileReader shape for large tensor materialization:
+// one disk-wide bounded pool, not one unbounded pool per safetensors blob.
+var readerBatchTokens = make(chan struct{}, readerBatchParallelism())
+
+func readerBatchParallelism() int {
+	return min(max(runtime.NumCPU()/2, 4), 16)
+}
+
+type fileIOReader struct {
```
Comment (jessegross) [inline x/mlxrunner/mlx/io_reader.go]: The mlx package should only carry the vtable adapter over an `io.ReaderAt` with `Size` and `Close`. Descriptions, batching, byte counting and the total belong in the runner, which already knows the manifest.

File: x/mlxrunner/runner.go [hunk 5]
```diff
@@ -278,9 +270,77 @@ func loadTensorsFromManifest(root *model.Root) (map[string]*mlx.Array, error) {
...
+		if n <= 0 {
+			return
+		}
+		done := completed.Add(n)
+		frac := float32(done) / float32(total)
+		if frac >= 1 {
+			// Keep 1.0 reserved for MarkReady. Reading tensor bytes is necessary
+			// but not sufficient: LoadWeights, final eval, and runner setup still
+			// have to complete before requests are safe to serve.
+			frac = math.Nextafter32(1, 0)
```
Comment (jessegross) [inline x/mlxrunner/runner.go]: Reads finish before `LoadWeights`, the eval and wiring, so the fraction legitimately reaches 1 before ready. The client's reset-on-change branch already grants a full stall window at that point, so the reservation and the `fullyLoaded` branch at client.go:151 can both go.

File: x/mlxrunner/runner.go [hunk 6]
```diff
@@ -141,7 +145,11 @@ func (r *Runner) loadModel(modelName string) (weights []*mlx.Array, err error) {
 		r.cache = newPrefixCache(slices.Concat(caches, draftCaches))
 		r.Sampler = sample.New(r.contextLength)
 		r.spec = newSpeculation(r, draftModel, caches, draftCaches)
-		r.grammarEngine = newGrammarEngine(logitsWidth(m), r.Tokenizer)
+		if vocabSize, e := root.VocabSize(); e != nil {
```
Comment (jessegross) [inline x/mlxrunner/runner.go]: This is unrelated to progress, and a config without the field now silently loses structured output. Keep the logits-width probe; if it misbehaved with the reader, that's a separate bug.

File: x/mlxrunner/server.go [hunk 7]
```diff
@@ -62,24 +64,35 @@ func Execute(args []string) error {
...
 	}
-	initialMemory, err := mlxthread.Call(context.Background(), worker, readMemory)
+	// Weight bytes from the manifest stand in until the model is loaded and the
+	// MLX worker is free to report real usage. Reporting zero instead would
+	// clobber the parent's own estimate, which it reads before the load ends.
+	modelManifest, err := manifest.LoadManifest(modelName)
 	if err != nil {
 		return err
 	}
+	loadingMemory := uint64(modelManifest.TotalTensorSize())
```
Comment (jessegross) [inline x/mlxrunner/server.go]: The parent only applies a status once it's ready, so this value and the manifest re-read are never consumed. Report memory from the cache as before.

File: x/mlxrunner/client.go [hunk 8]
```diff
@@ -446,39 +492,65 @@ func (c *Client) Pid() int {
...
+// Ping implements llm.LlamaServer.
+func (c *Client) Ping(ctx context.Context) error {
+	status, err := c.getServerStatus(ctx)
+	if err != nil {
 		return err
 	}
 
-	c.contextLength.Store(int64(c.reportedContextLength(status.ContextLength)))
-	c.memory.Store(status.Memory)
+	if status.Status != llm.ServerStatusReady {
```
Comment (jessegross) [inline x/mlxrunner/client.go]: `needsReload` blocks on the runner mutex until the load returns, so the scheduler can't read a partial `MemorySize`. Should Ping stay healthy while loading, as llama-server's does?

**Review by jessegross — COMMENTED** [review]
The progress mechanism should be just the Go reader adding bytes to a counter, the manifest total, and a ready flag, with `/v1/status` reporting the fraction. The native loader path, the `Nextafter32` reservation and `fullyLoaded` branch, the `vocab_size` parser, the loading-phase memory value, and the reader's error state are all guarding cases the parent never reaches or are unrelated to progress, and they can be removed. Details inline.
